src/workers/python/init.py

```diff
@@ -9,19 +11,79 @@
 await micropip.install("python_runner")
 import python_runner
 
+# Install modules asynchronously to use when the user needs them
 ft = micropip.install("friendly_traceback")
 jedi_install = micropip.install("jedi")
-# Otherwise `import matplotlib` fails while assuming a browser backend
-os.environ["MPLBACKEND"] = "AGG"
 
-# Code is executed in a worker with less resources than ful environment
-sys.setrecursionlimit(500)
+SYS_RECURSION_LIMIT = 500
 
 # Global Papyros instance
 papyros = None
 
+class Papyros(python_runner.PyodideRunner):
+    def __init__(
+        self,
+        *,
+        callback=None,
+        limit=SYS_RECURSION_LIMIT
+    ):
+        super().__init__()
+        self.limit = limit
+        self.override_globals()
+        self.set_event_callback(callback)
+
+    def set_event_callback(self, event_callback):
+        if event_callback is None:
+            raise ValueError("Event callback must not be None")
+
+        def runner_callback(event_type, data):
+            def cb(typ, dat, **kwargs):
+                return event_callback(to_js(dict(type=typ, data=dat, **kwargs)))
+
+            # Translate python_runner events to papyros events
+            if event_type == "output":
+                for part in data["parts"]:
+                    typ = part["type"]
+                    if typ in ["stderr", "traceback", "syntax_error"]:
+                        cb("error", part["text"], contentType="text/json")
+                    elif typ == "stdout":
+                        cb("output", part["text"], contentType="text/plain")
+                    elif typ == "img":
+                        cb("output", part["text"], contentType=part["contentType"])
+                    elif typ in ["input", "input_prompt"]:
+                        continue
+                    else:
+                        raise ValueError(f"Unknown output part type {typ}")
+            elif event_type == "input":
+                return cb("input", data["prompt"], contentType="text/plain")
+            elif event_type == "sleep":
+                return cb("sleep", str(data["seconds"]*1000), contentType="text/integer")
+            else:
+                raise ValueError(f"Unknown event type {event_type}")
+
+        self.set_callback(runner_callback)
+
+    def set_file(self, filename, code):
+        self.filename = os.path.normcase(os.path.abspath(filename))
+        with open(self.filename, "w") as f:
+            f.write(code)
```

Comment:
This should really happen in python_runner, I've noticed not having a real file causing bugs in futurecoder.

What's the normcase for?

Reply:
The normcase is what Pdb uses to resolve filenames, so I figured I would use their method to construct filenames so that every module would find and use the same file.

Reply:
Is this then a feature you will add to python_runner, or what is our plan with this? The current python_runner API always works with source code or code objects, which is the reason why I added this myself.

Reply:
I think this exact code should be in pre_run next to the linecache stuff, but with errors caught in case python_runner is used in places where writing to the filesystem is not an option.

Reply:
resolved